app/code/Magento/MediaGallery/Model/ResourceModel/Keyword/SaveAssetLinks.php

```diff
@@ -56,27 +66,124 @@ public function __construct(
     public function execute(int $assetId, array $keywordIds): void
     {
         try {
-            $values = [];
-            foreach ($keywordIds as $keywordId) {
-                $values[] = [$assetId, $keywordId];
+            $currentKeywordIds = $this->getCurrentKeywordIds($assetId);
+
+            $obsoleteKeywordIds = array_diff($currentKeywordIds, $keywordIds);
+            $newKeywordIds = array_diff($keywordIds, $currentKeywordIds);
+
+            $this->deleteAssetKeywords($assetId, $obsoleteKeywordIds);
+            $this->insertAssetKeywords($assetId, $newKeywordIds);
+
+        } catch (\Exception $exception) {
+            $this->logger->critical($exception);
+            throw new CouldNotSaveException(
+                __('Could not process asset keyword links'),
+                $exception
+            );
+        }
+    }
+
+    /**
+     * Save new asset keyword links
+     *
+     * @param int $assetId
+     * @param int[] $keywordIds
+     * @throws CouldNotSaveException
+     */
+    private function insertAssetKeywords(int $assetId, array $keywordIds): void
+    {
+        try {
+            if (!empty($keywordIds)) {
```

Comment:
I'd revert the conditions to decrease the nesting level (this can be applied to other such conditions in this PR as well). Also this check can be moved out of try-catch block

if (empty($keywordIds)) {
    return;
}